core/include/thread.h

```diff
@@ -526,6 +526,81 @@ static inline bool thread_is_active(const thread_t *thread)
  */
 const char *thread_state_to_string(thread_status_t state);
 
+/**
+ * Get start address (lowest) of a thread's stack.
+ *
+ * @param   thread thread to work on
+ * @returns current stack pointer, or NULL if not available
+ */
+static inline void *thread_get_stackstart(const thread_t *thread)
+{
+#if defined(DEVELHELP) || defined(SCHED_TEST_STACK) \
+    || defined(MODULE_MPU_STACK_GUARD)
+    return thread->stack_start;
+#else
+    (void)thread;
+    return NULL;
+#endif
+}
+
+/**
+ * Get stored Stack Pointer of thread.
+ *
+ * *Only provides meaningful value if the thread is not currently running!*.
+ *
+ * @param   thread thread to work on
+ * @returns current stack pointer
+ */
+static inline void *thread_get_sp(const thread_t *thread)
```

Comment:
I wonder if those should return `const void *` so people don't get funky ideas

Reply:
should it? I don't mind, it could mean casting down the line (e.g., in ps()).

Reply:
Where's the "grinning smiley with devil's horns" react emoji when you need it? >;-D

Reply:
Does `ps` require non-const access to the thread stack? :fearful:

Reply:
`((void *)thread_get_sp(thread))[123] = 0xD1ED1E;`

ah wait, stack grows downwards. :/

Reply:
If the thread is reliably stopped, couldn't it even be used meaningfully by resetting the thread's high water mark? (Then the non-const would be justified, although the constness of a pointer that arbitrary arithmetic is later done to has little meaning anyway).

Reply:
Like, for resetting the stack measurement? That could be useful.

Do we need const protection from devs that want to intentionally abuse a given stack pointer? `int foo; void *sp = &foo;` goes a long way for those inclined.

Reply:
If you can think of use cases where `void *` is needed than that's fine.
I just try to use `const` whenever possible.